### PR TITLE
packaging: add testing environment

### PR DESCRIPTION
**Pull request by Choraden**
Packaging testing environment for forwarder installation. The environment is based on an image with systemd support and is meant to be used with Podman. It provides a Dockerfile to build the container image and a Makefile for easier container management.

### CODE REVIEW

File: packaging/systemd/forwarder.service [hunk 1]
```diff
@@ -4,8 +4,8 @@ After=network-online.target
 
 [Service]
 Type=simple
-User=nobody
-Group=nobody
+User=forwarder-sa
```
Comment (mmatczuk) [inline packaging/systemd/forwarder.service:7]: can we use user forwarder?

File: packaging/test/README.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# Packaging Testing Environment
+
+This is a packaging testing environment for testing your service's installation within a container. The environment is based on an image with systemd support and is meant to be used with Podman. It provides a Dockerfile to build the container image and a Makefile for easier container management.
```
Comment (mmatczuk) [inline packaging/test/README.md:3]: Use sentence per line.

File: packaging/test/README.md [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+## Prerequisites
+
+- [Podman](https://podman.io/) installed on your system
+
+## Getting Started
+
+- build packages: `./bin/goreleaser release --skip-publish --snapshot --skip-docker --clean` from forwarder root directory
```
Comment (mmatczuk) [inline packaging/test/README.md:11]: There is a make target for that if not there should be.

File: local/README.md [hunk 4]
```diff
@@ -0,0 +1,19 @@
+# Packaging Testing Environment
+
+This is a packaging testing environment for testing forwarder's installation.
```
Comment (mmatczuk) [inline local/README.md:3]: Forwarder's

Comment (mmatczuk) [reply local/README.md:3]: Forwarder

File: local/README.md [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+The environment is based on an image with systemd support and is meant to be used with Podman.
```
Comment (mmatczuk) [inline local/README.md:4]: based on Docker image

File: local/README.md [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+It provides a Dockerfile to build the container image and a Makefile for easier container management.
+
+## Prerequisites
+
+- [Podman](https://podman.io/) installed on your system
+
+## Getting Started
+
+- build packages: ` from forwarder root directory
```
Comment (mmatczuk) [inline local/README.md:13]: Quote?

File: local/rpm/Makefile [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+.PHONY: build
+build: package
+	@cp ../../dist/forwarder*_linux.aarch64.rpm .
+	@mv forwarder*_linux.aarch64.rpm forwarder.aarch64.rpm
+	@podman build -t forwarder-rpm .
+	@rm forwarder.aarch64.rpm
+
+.PHONY: package
+package:
+	@cd ../../ && ./bin/goreleaser release --skip-publish --snapshot --skip-docker --clean
```
Comment (mmatczuk) [inline local/rpm/Makefile:15]: This can build only linux images.

File: local/rpm/Makefile [hunk 8]
```diff
@@ -0,0 +1,23 @@
+.PHONY: run
+run:
+	@podman run -ti -p 3128:3128 -d --name forwarder-rpm forwarder-rpm
+	@podman exec -ti forwarder-rpm /bin/sh
+
+.PHONY: build
+build: package
+	@cp ../../dist/forwarder*_linux.aarch64.rpm .
```
Comment (mmatczuk) [inline local/rpm/Makefile:8]: Use variable for dist location.

File: local/debian/Makefile [hunk 9]
```diff
@@ -0,0 +1,27 @@
+IMG ?= saucelabs/forwarder:debian
```
Comment (mmatczuk) [inline local/debian/Makefile:1]: Let's call it forwarder-testing or something

Comment (mmatczuk) [reply local/debian/Makefile:1]: Scratch that, it's debian-systemd ofc

File: local/rpm/Makefile (hunk 7 above)
Comment (Choraden) [reply local/rpm/Makefile]: How do I build nfpms only?

File: local/debian/Makefile [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+package:
+	@cd ../../ && ./bin/goreleaser release --snapshot --skip-docker --clean
+	@cp $(DIST) forwarder.deb
+
+.PHONY: run
+run:
+	@podman run -p 3128:3128 -d -v ./forwarder.deb:/forwarder.deb --name $(CONTAINER) $(IMG)
+	@podman exec $(CONTAINER) dpkg -i /forwarder.deb
+	@podman exec $(CONTAINER) systemctl start forwarder
+	@podman exec -ti $(CONTAINER) /bin/sh
```
Comment (mmatczuk) [inline local/debian/Makefile:19]: Can we have it like run -it so that there is no need for stop?

File: local/debian/Makefile [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+CONTAINER ?= forwarder-debian
+
+.PHONY: build
+build:
+	@podman build -t $(IMG) .
+
+.PHONY: package
+package: DIST ?= ../../dist/forwarder*linux_arm64.deb
+package:
+	@cd ../../ && ./bin/goreleaser release --snapshot --skip-docker --clean
```
Comment (mmatczuk) [inline local/debian/Makefile:11]: I was thinking maybe add soft link to git?

Comment (mmatczuk) [reply local/debian/Makefile:11]: Could link this work as a volume?

File: local/rpm/Makefile (hunk 7 above)
Comment (mmatczuk) [reply local/rpm/Makefile]: How about --single-target?

Comment (mmatczuk) [reply local/rpm/Makefile]: This works with build only... https://github.com/goreleaser/goreleaser/issues/3219

File: local/debian/Makefile (hunk 10 above)
Comment (Choraden) [reply local/debian/Makefile:19]: The entrypoints for systemd to work must be `/sbin/init`, or `/lib/systemd/systemd`.
```
If the command you running inside of the container is systemd or init, podman will setup tmpfs mount points in the following directories...
```

Running init requires user to log in and some additional setup, thus I made this little workaround.

File: local/debian/Makefile (hunk 11 above)
Comment (Choraden) [reply local/debian/Makefile:11]: I found the copy technique convenient to use, as it allows us to use pattern matching `*` (SNAPSHOT releases have commit hash in the file name). I doubt you can mount a volume without the knowledge of full file name.

File: local/debian/Makefile (hunk 10 above)
Comment (Choraden) [reply local/debian/Makefile:19]: How about we remove the container once the `podman exec -ti $(CONTAINER) /bin/sh` exit?

File: local/debian/Makefile (hunk 11 above)
Comment (mmatczuk) [reply local/debian/Makefile:11]: I see.

Comment (mmatczuk) [reply local/debian/Makefile:11]: The full file name is the name of the link. The question is can it read links.

Comment (Choraden) [reply local/debian/Makefile:11]: We can link it the same way we copy - with the `*`. 
```
.PHONY: package
package: DIST ?= ../../dist/forwarder*linux_arm64.deb
package:
	@cd ../../ && ./bin/goreleaser release --snapshot --skip-docker --clean
	@ln -s $(DIST) forwarder.deb
```

Comment (Choraden) [reply local/debian/Makefile:11]: Never mind. Soft links seem to be messing something with the Podman build 
```
make build  
Error: invalid symlink "/var/tmp/libpod_builder4117844942/build/forwarder.deb" -> "../../dist/forwarder_1.1-rc1-SNAPSHOT-d9b0b83.linux_arm64.deb"
gmake: *** [Makefile:6: build] Error 125
```

File: .gitignore [hunk 12]
```diff
@@ -9,6 +9,10 @@
 *.dylib
 .direnv
 
+# Linux packages
```
Comment (mmatczuk) [inline .gitignore:12]: Wrong dir.

File: local/debian/Makefile [hunk 13]
```diff
@@ -0,0 +1,24 @@
+IMG ?= saucelabs/forwarder:debian-systemd
+CONTAINER ?= forwarder-debian
```
Comment (mmatczuk) [inline local/debian/Makefile:2]: How about

IMG = debian-systemd:bullseye
CONTAINER = saucelabs/forwarder-testing-debian

File: .gitignore (hunk 12 above)
Comment (Choraden) [reply .gitignore:12]: Do you suggest to ignore it only in `/local/`? I see no circumstances in which we would commit a .deb package, but ok.

Comment (mmatczuk) [reply .gitignore:12]: It's out of context here.

File: local/debian/Makefile (hunk 13 above)
Comment (Choraden) [reply local/debian/Makefile:2]: CONTAINER could be `forwarder-testing-debian` as it must match ` [a-zA-Z0-9][a-zA-Z0-9_.-]*:`

**Comment by Choraden** [comment]
@mmatczuk Makefiles were rewritten into run scripts.

**Comment by mmatczuk** [comment]
Notes on bash scripts
* Copyright
* better shebang `#!/usr/bin/env bash`
* fail fast, add `set -eu -o pipefail`

Otherwise LGTM

**Pull request merged**